update node controller to use shared pod informer

sjenning opened the pull request:
continuing work from #24470 and #23575

ncdc commented:
cc @kubernetes/rh-cluster-infra @smarterclayton @deads2k @liggitt

deads2k commented:
> This PR is not for the master branch but does not have the cherrypick-approved label. Adding the do-not-merge label.

@eparis I feel like the bot is trying to tell me something, but I don't know what.

eparis commented:
Its trying to tell you the github itself had major problems and the bot lost it's mind.

davidopp commented:
not sure who is the best reviewer -- @lavalamp ? @wojtek-t ? @bprashanth ? Assigning to @wojtek-t for now.

kargakis commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -355,7 +355,7 @@ func StartControllers(s *options.CMServer, kubeClient *client.Client, kubeconfig
 		if containsResource(resources, "deployments") {
 			glog.Infof("Starting deployment controller")
-			go deployment.NewDeploymentController(clientset.NewForConfigOrDie(restclient.AddUserAgent(kubeconfig, "deployment-controller")), ResyncPeriod(s)).
+			go deployment.NewDeploymentController(podInformer, clientset.NewForConfigOrDie(restclient.AddUserAgent(kubeconfig, "deployment-controller")), ResyncPeriod(s)).
```
Do not update the deployment controller to share the pod cache just yet because it mutates its pod cache! I've opened https://github.com/kubernetes/kubernetes/issues/27687

kargakis commented:
Make sure the rest of the controllers don't mutate too

smarterclayton commented:
is this still on track for 1.4?

sjenning commented:
@smarterclayton i need to pick this back up.  Apparently the deployment controller mutates the cache so I'll have to drop that for now.  This PR will only convert the node controller in that case.

smarterclayton commented:
@karkagis re: cache mutating deployment controller

On Wed, Jul 20, 2016 at 12:45 PM, Seth Jennings <notifications@github.com>
wrote:

> @smarterclayton <https://github.com/smarterclayton> i need to pick this
> back up. Apparently the deployment controller mutates the cache so I'll
> have to drop that for now. This PR will only convert the node controller in
> that case.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24841#issuecomment-234008148>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_pyn6D9kR7KZ7vZyTMGOSTLIO17XFks5qXlCbgaJpZM4IQjMm>
> .
>

quinton-hoole commented:
@mwielgus FYI, as this intersects with some stuff he's doing on Federation.

wojtek-t commented:
@quinton-hoole - I don't think so; this is something different

sjenning committed:
refactor maybeDeleteTerminatingPod

sjenning committed:
node controller use shared pod informer

sjenning commented:
Ok finally updated this.  Dropped the deployment controller (see above why), rebased, refactored a change that introduced an incompatibility with the node controller using the shared informer.

kargakis commented:
@deads2k is the cache mutation detector enabled for Kube?

wojtek-t commented:
@kargakis what is "cache mutation detector"?

kargakis commented:
> @kargakis what is "cache mutation detector"?

https://github.com/kubernetes/kubernetes/pull/27784

deads2k commented:
> @kargakis what is "cache mutation detector"?

It allows detection of cache mutations based on an env var.  It's not cheap to run, but its useful when running e2e style tests to find people who are messing with the cache.  I'd really like to see it (or something that does the same job) merged.

wojtek-t commented:
LGTM
